src/openssl_auth.cpp

```diff
@@ -238,8 +245,66 @@ bool s3fs_HMAC256(const void* key, size_t keylen, const unsigned char* data, siz
     return s3fs_HMAC_RAW(key, keylen, data, datalen, digest, digestlen, true);
 }
 
+#ifdef USE_OPENSSL_30
 //-------------------------------------------------------------------
-// Utility Function for MD5
+// Utility Function for MD5 (OpenSSL >= 3.0)
+//-------------------------------------------------------------------
+// [NOTE]
+// OpenSSL 3.0 deprecated the MD5_*** low-level encryption functions,
+// so we should use the high-level EVP API instead.
+//
+size_t get_md5_digest_length()
+{
+    return EVP_MD_size(EVP_md5());
+}
+
+unsigned char* s3fs_md5_fd(int fd, off_t start, off_t size)
+{
+    EVP_MD_CTX*    mdctx;
+    unsigned char* md5_digest;
+    unsigned int   md5_digest_len = get_md5_digest_length();
+    off_t          bytes;
+
+    if(-1 == size){
+        struct stat st;
+        if(-1 == fstat(fd, &st)){
+            return NULL;
+        }
+        size = st.st_size;
+    }
+
+    // instead of MD5_Init
+    mdctx = EVP_MD_CTX_new();
+    EVP_DigestInit_ex(mdctx, EVP_md5(), NULL);
+
+    for(off_t total = 0; total < size; total += bytes){
+        const off_t len = 512;
+        char        buf[len];
+        bytes = len < (size - total) ? len : (size - total);
+        bytes = pread(fd, buf, bytes, start + total);
+        if(0 == bytes){
+            // end of file
+            break;
+        }else if(-1 == bytes){
+            // error
+            S3FS_PRN_ERR("file read error(%d)", errno);
+            return NULL;
```

Comment:
Should this call `EVP_MD_CTX_free`?

Reply:
Sorry for my fault, it is fixed and updated soon.